### PR TITLE
PERF: join/merge on subset of MultiIndex

### PR DESCRIPTION
**Pull request by lukemanley**
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/v1.6.0.rst` file if fixing a bug or adding a new feature.

Existing code passes a range to algos.take_nd. Passing an ndarray is faster. 

```
       before           after         ratio
     [a712c501]       [00034d83]
     <main>           <multiindex-join-subset>
-    45.8±0.7ms       26.0±0.8ms      0.57  join_merge.JoinMultiindexSubset.time_join_multiindex_subset
```

### CODE REVIEW

File: pandas/core/reshape/merge.py [hunk 1]
```diff
@@ -1678,21 +1678,19 @@ def _convert_to_multiindex(index: Index) -> MultiIndex:
...
-    if rindexer is None:
-        rindexer = range(right.size)
-
     # Iterate through the levels that must be restored
     for dropped_level_name in dropped_level_names:
+        # lindexer and rindexer hold the indexes where the join occurred
+        # for left and right respectively. If left/right is None then
+        # the join occurred on all indices of left/right
         if dropped_level_name in left.names:
+            if lindexer is None:
```
Comment (phofl) [inline pandas/core/reshape/merge.py:1687]: Why are you pulling this inside the loop? This is harder to read than before

Comment (lukemanley) [reply pandas/core/reshape/merge.py]: Agreed. We can actually avoid calling take_nd when the indexers are None as that means "take everything". I made another commit which further improves times and should be clearer.

```
       before           after         ratio
     [a712c501]       [ad3f42b5]
     <main>           <multiindex-join-subset>
-    43.8±1ms         10.5±1ms        0.24  join_merge.JoinMultiindexSubset.time_join_multiindex_subset
```

File: pandas/core/reshape/merge.py [hunk 2]
```diff
@@ -1679,11 +1679,6 @@ def _convert_to_multiindex(index: Index) -> MultiIndex:
     # lindexer and rindexer hold the indexes where the join occurred
     # for left and right respectively. If left/right is None then
     # the join occurred on all indices of left/right
-    if lindexer is None:
```
Comment (mroeschke) [inline pandas/core/reshape/merge.py:1682]: Is the above comment relevant/can be removed?

Comment (lukemanley) [reply pandas/core/reshape/merge.py:1721]: I've removed the comment now that we're not creating a range of all indices.

**Review by mroeschke — APPROVED** [review]
LGTM. Merge when ready @phofl

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx @lukemanley

**Pull request merged**